test/jdk/java/nio/channels/DatagramChannel/StressNativeSignal.java

```diff
@@ -39,55 +45,78 @@ public class StressNativeSignal {
         try {
             serverSocketThread = new ServerSocketThread();
             serverSocketThread.start();
+        } catch (Exception z) {
+            System.err.println("failed to create and start a ServerSocketThread");
+            z.printStackTrace();
+        }
 
+        try {
             udpThread = new UDPThread();
             udpThread.start();
         } catch (Exception z) {
+            System.err.println("failed to create and start a UDPThread");
             z.printStackTrace();
         }
     }
 
     public static void main(String[] args) throws Throwable {
         StressNativeSignal test = new StressNativeSignal();
-        try {
-            Thread.sleep(3000);
-        } catch (Exception z) {
-            z.printStackTrace(System.err);
-        }
-
+        test.waitForTestThreadsToStart();
         test.shutdown();
     }
 
     public void shutdown() {
-        udpThread.terminate();
-        try {
-            udpThread.join();
-        } catch (Exception z) {
-            z.printStackTrace(System.err);
+        if ((udpThread != null) && udpThread.isAlive()) {
+            udpThread.terminate();
+            try {
+                udpThread.join();
+            } catch (Exception z) {
+                z.printStackTrace(System.err);
+            }
+        }
+        if ((serverSocketThread != null) && (serverSocketThread.isAlive())) {
+            serverSocketThread.terminate();
+            try {
+                serverSocketThread.join();
+            } catch (Exception z) {
+                z.printStackTrace(System.err);
+            }
         }
+    }
 
-        serverSocketThread.terminate();
-        try {
-            serverSocketThread.join();
-        } catch (Exception z) {
-            z.printStackTrace(System.err);
+    public void waitForTestThreadsToStart() {
+        if ((udpThread != null) && udpThread.isAlive()) {
+            udpThread.waitTestThreadStart();
+        }
+        if ((serverSocketThread != null) && (serverSocketThread.isAlive())) {
+            serverSocketThread.waitTestThreadStart();
         }
     }
 
     public class ServerSocketThread extends Thread {
         private volatile boolean shouldTerminate;
         private ServerSocket socket;
+        private final CountDownLatch threadStarted = new CountDownLatch(1);
+
+        public ServerSocketThread () throws Exception {
+
+            socket = new ServerSocket(1122);
```

Comment:
Agreed.

Reply:
I noted this in my introduction. It could be changed to wildcard, but it may lead to intermittent failures on macOS which is usually rectified by explicitly using loopback address to avoid some idiosyncratic port allocation.

It is note, also, that the UDP port is hardcoded to one from the IANA reserve port range, but uses an unassigned port in that range.

Reply:
Not wildcard address, as it is already on wildcard but hardcoded port  change  to OS ephemeral port

Reply:
That sounds fine then.